### PR TITLE
Fix column metadata getMaxValue NPE bug and expose maxNumMultiValues

### PR DESCRIPTION
**Pull request by mqliang**
## Description

This PR:

* fix a bug of `metadata.getMaxValue()` may return a null pointer, see the function here:
https://github.com/apache/pinot/blob/77a706996099f9bb44b90ad506f5205b3c4d7a42/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/SegmentColumnarIndexCreator.java#L667-L675

When the maxValue/minValue exceed METADATA_PROPERTY_LENGTH_LIMIT, SegmentColumnarIndexCreator will not set minValue/minValue. In this case, `metadata.getMaxValue()' will return NULL.

* expose maxNumMultiValues information for MV columns
* add integration test case for MV columns and column name encoding

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

**Review by Jackie-Jiang — COMMENTED** [review]
Shall we add some tests to verify the expected behavior and guard it from regression?

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java [hunk 1]
```diff
@@ -210,14 +212,15 @@ public String getSegmentMetadata(
...
             } else if (storedDataType == DataType.STRING || storedDataType == DataType.BYTES) {
               // For type of variable width (String, Bytes), if it's stored using raw bytes, set the columnLength as
-              // the length
-              // of the max value.
-              columnLength = ((String) columnMetadata.getMaxValue()).getBytes(StandardCharsets.UTF_8).length;
+              // the length of the max value.
+              ByteArray maxValueByteArray =
+                  storedDataType == DataType.BYTES ? ((ByteArray) columnMetadata.getMaxValue())
+                      : BytesUtils.toByteArray((String) columnMetadata.getMaxValue());
+              columnLength = maxValueByteArray.length();
```
Comment (walterddr) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java:223]: could you add test for this?

**Review by walterddr — COMMENTED** [review]
+1 to @Jackie-Jiang 's comment. also bug1 seems to be an improvement right? improving the inefficiency of encode/decode.

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java (hunk 1 above)
Comment (siddharthteotia) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java:223]: This code change does not look correct to me. AFAIK, we want to fix couple of problems

- Use of * to allow fetching metadata for all columns instead of individual column. Looks like this has been fixed by encoding * ? Please add tests for the same

- A weird scenario where the use of particular column name (internal) fails the API call for some reason. That column type is STRING and stored type is also STRING (UTF-8 bytes essentially). I don't think this is related to the handling of BYTES / BYTE_ARRAY. While column of data types / stored types as BYTES and BYTE_ARRAY should also be handled, the problem we wanted to fix does not seem to be related to it or at least there is no evidence from debugging. I suggest debugging this more to understand the root cause and then come back on this part of the change

Comment (siddharthteotia) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java]: Discussed offline with @mqliang 

As suggested in the previous comment, let's not fix the BYTES problem in this PR. We can investigate and fix that separately. The url encoding for (with and without *) and that corner case problem for a particular problem should be fixed (which @mqliang  found out is due to min/max value being null and server throwing NPE). 
Let's make sure to add tests for with and without encoded query string -- everything after columns= can be encoded. I think in this PR, controller is re-encoding. We should remove that

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java [hunk 2]
```diff
@@ -210,21 +211,22 @@ public String getSegmentMetadata(
...
+              columnLength = maxString == null ? 0 : maxString.getBytes(StandardCharsets.UTF_8).length;
             } else {
               // For type of STRUCT, MAP, LIST, set the columnLength as DEFAULT_MAX_LENGTH (512).
               columnLength = FieldSpec.DEFAULT_MAX_LENGTH;
             }
             int columnCardinality = segmentMetadata.getColumnMetadataMap().get(column).getCardinality();
+            int maxNumMultiValues = segmentMetadata.getColumnMetadataMap().get(column).getMaxNumberOfMultiValues();
             columnLengthMap.merge(column, (double) columnLength, Double::sum);
             columnCardinalityMap.merge(column, (double) columnCardinality, Double::sum);
+            maxNumMultiValuesMap.merge(column, (double) maxNumMultiValues, Double::sum);
```
Comment (siddharthteotia) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java:229]: Can we avoid populating this map for single valued column since maxNumMultiValues is only applicable to MV columns ? The other option is to make sure it is either 0 or -1 for SV columns

**Comment by siddharthteotia** [comment]
@mqliang can you please update the PR description to correctly reflect the actual changes that we finally agreed to do in this PR ? 

Let's remove the stuff related to byte because it turned out that was not the issue and let's describe the reason for the NPE issue we were seeing due to METADATA_PROPERTY_LENGTH_LIMIT (it is described in our internal ticket).

Regarding the addition of maxNumMultiValues, can we ensure in tests that it is non-zero only for columns that are multi-value and 0 or -1 for single value columns as it will never be used for them

**Review by siddharthteotia — APPROVED** [review]
